### ISSUE TITLE
Add download link in admin area when folder is not accessed with http

### ISSUE THREAD

**Issue by abeel** [easydigitaldownloads/Easy-Digital-Downloads#3002]
for each product let the admin to download the file if the edd download folder isnot accessed with http (outside the site folder) 

File URL : /var/downloads/download/edd/2015/01/download.zip

**Comment by pippinsplugins** [comment]
We've had this requested a few times and I think it's a good idea.

**Comment by cklosowski** [comment]
I must be missing something. Can you e plain the use case @pippinsplugins?

**Comment by pippinsplugins** [comment]
If you copy and paste the file URLs from the edit screen of a Download product, you will be given a "Permission denied" error message due to the htaccess file.

Imagine a site using Frontend Submissions. One of the ways that site admins review files is by downloading them. With the current system, admins are required to download them via FTP to view the files.

**Comment by abeel** [comment]
@pippinsplugins +1

**Comment by cklosowski** [comment]
Ah. Perfect, thanks. :100:

**Comment by ghost1227** [comment]
What do we want this to look like?

**Comment by pippinsplugins** [comment]
Take a look at WPeC, they have a pretty nice system: 
![screen shot 2015-01-27 at 3 09 45 pm](https://cloud.githubusercontent.com/assets/1034109/5927167/954e0194-a636-11e4-8467-f89ac5e9588e.png)

**Comment by ghost1227** [comment]
Do we _only_ want it to show if the downloads aren't accessible view http, or always show it?

**Comment by pippinsplugins** [comment]
Show regardless.

**Comment by cklosowski** [comment]
![image](https://cloud.githubusercontent.com/assets/1390124/8167495/8d58504a-1353-11e5-9663-158145bf1d9d.png)
@pippinsplugins You think this should do it? Could use some testing for Dropbox, Amazon, etc.

**Comment by cklosowski** [comment]
Also adding logging and Nonce verification. Now when an Admin downloads the file:
![image](https://cloud.githubusercontent.com/assets/1390124/8167981/99a0a764-1356-11e5-93b7-0dfad166e582.png)

**Comment by pippinsplugins** [comment]
@cklosowski A few things:

1. I don't think this should be a separate meta box. It should somehow be incorporated into the existing Download Files metabox, otherwise it creates a disconnect and will probably cause confusion. "Why are there files here and there?"

2. If adding a download processor for admins, we should extract as much out of the existing ones so that both admin/customer methods share code instead of duplicating it.

**Comment by cklosowski** [comment]
@pippinsplugins 
1) Totally disagree, I think it offers more confusion if it's just below in the same metabox. Plus it starts making that a 'God Metabox', just WAY too much going on in it.
2) I was pretty conservative in what I ripped out, I'll try to rip a few more things into a function.

**Comment by pippinsplugins** [comment]
1). Ignoring possible issues with making the metabox too large, can you give me your reasons for feeling that a separate box is less confusing?

**Comment by cklosowski** [comment]
1) It has a separate permission than the regular metabox
2) It makes it possible to hide it, if you aren't running FES or the use case isn't valid for you
3) We don't show it if it's part of bundles, this means if it's part of the other metabox, you have too add more html that may (or may not) ever display and then ajax must update it. If it's a bundle, we just don't need to show it.

**Comment by cklosowski** [comment]
The other option we have is something like this:
![image](https://cloud.githubusercontent.com/assets/1390124/8169410/3d8d78e6-135e-11e5-8ca0-812ded2ea41a.png)

**Comment by pippinsplugins** [comment]
@easydigitaldownloads/core-devs What do you think guys? Which option do you prefer?

**Comment by SDavisMedia** [comment]
Uggghhh I **really** like the separate metabox and the reasons for having it. But I think having the actions column in the existing metabox makes more sense for those who will actually use it.

**Comment by cklosowski** [comment]
I read that as Sean likes both ;-)

**Comment by mindctrl** [comment]
Option 1. Less change. Nice icons, though they might be too close and someone could accidentally delete instead of download.

With space being an issue in this box, maybe we should consider shortening the URL box or coming up with an interesting way to hide the URL by default and show it with a click or something. It rarely, if ever, shows the whole URL anyway.

**Comment by pippinsplugins** [comment]
Is no one else concerned with the possibility of people getting confused y two meta boxes? Even as a power user, having data about the file downloads in two places seems super confusing to me.

**Comment by SDavisMedia** [comment]
@pippinsplugins I'm concerned with it. That's why I chose Camera 1.

To avoid getting too cluttered in the metabox, assuming we go with Camera 1, consider combining the File Name and File URL columns. Imagine this:

One column called **File Details** that is the width of the current File URL column and is positioned all the way left. Stretch each row vertically just a liiiittle bit and add (in smaller, bold text) the name of the file **above** the File URL field. Click-to-edit the file name (maybe a pencil icon nearby). 

That frees up 20-25% horizontal space in the metabox and allows us to use Camera 1, which again I think makes more sense.

**Comment by pippinsplugins** [comment]
File name and URL cannot be combined. People manually customize the file name some times.

**Comment by cklosowski** [comment]
We can add the table of file details to the other meta box however I'm getting concerned with the maintainability of that metl box from a code perspective

**Comment by SDavisMedia** [comment]
I already said click-to-edit the file name in my last reply.

![2015-06-15 at 3 38 pm](https://cloud.githubusercontent.com/assets/2359131/8170227/d118354a-1374-11e5-9e42-530063de1029.png)

**Comment by cklosowski** [comment]
I think the download information is important to show. Let me try something, against my opinion, but adding some of this information to the same metabox.

**Comment by cklosowski** [comment]
Punting for further discussion as there is clearly some disagreements in what this should be, and how it should function.

**Comment by cklosowski** [comment]
Just something to bring up, if we can get away from using Tables for the file list, this will open up a lot of possibilities in what data we can display via show/hide links. Right now it's restricted to a `<tr>` which isn't really easy to expand vertically.

We could convert the repeatable row to be divs, and that would help us go a long way with expanding it's extensibility. The issue however is a few hooks into the repeatable row that are expecting the hook to be within a `tr`

**Comment by rubengc** [comment]
A possible solutions could be add row actions under file URL column (similar to wordpress row actions) with upload/download/clear actions

**Comment by cklosowski** [comment]
So with the work that @SDavisMedia did on #2704, I think we can compleatly rework the repeatable file fields to work the same way, with an 'Advanced Options' (or some other wording like "File Information") link that drops down an area.

**Comment by cklosowski** [comment]
Ok, I think I have this reworked using the repeatable fields 'info' sections, as well as logging file downloads as admins (wether it's a local or remote file).

Here's a screenshot of the implementation, and like the 'Advanced Settings' line for variable pricing, there is a link to 'view file information'.
![screen shot 2017-08-28 at 12 22 05 pm](https://user-images.githubusercontent.com/1390124/29795989-b12ff81c-8c03-11e7-84bb-7f3cfbe5c418.png)



This is ready for some testing.

**Comment by chriscct7** [comment]
I would change File Location where the value is "Local" to "Uploaded" or "WordPress" or something. "Local" could be interpreted as it's on the computer of the person whose currently logged in.

I personally think "WordPress" for on-site and "External Site" for off-site would be less confusing. Then Amazon s3 for instance could hook in and change the label for their files for "Amazon S3"

Otherwise :+1:

**Comment by cklosowski** [comment]
Thanks. Will look at options. I think WordPress is a bit too general
because it could be an FTPd file. So maybe just the site URL?

**Comment by chriscct7** [comment]
Well a normal user isn't going to think FTP vs WP Media will matter as to where the site is. Is the point to say it's not in wp-content if it's not?

If so maybe:
- FTP Folder
- Media Library (or Upload Folder)
- External Site
- Amazon s3 (etc)

**Comment by cklosowski** [comment]
@chriscct7 Since the file input field supports relative URL, full file path, and media library, I would think just stating it's delivered via your site `site_url` would be enough here. Then we can add a filter for things like S3 and Dropbox to change that in a future update.

**Comment by chriscct7** [comment]
That's fair, though the site urls can get really long. I still think just "Website" would be fine. The file location regardless of if its ftp'd or whatever is still served from this website.

**Comment by mindctrl** [comment]
Why do you need to show `File Location`? The File URL basically reveals that already.

**Comment by chriscct7** [comment]
I agree with John for the most part, though it could be nice for Amazon s3 users to see that a file is uploaded to s3 or not, particularly if someone just installed S3 and is trying to quickly find files they need to re-upload to s3.

**Comment by mindctrl** [comment]
Since the heading above is `File Information`, repeating "File" on every item below it (File Location, File Size, etc) feels redundant.

**Comment by mindctrl** [comment]
Doesn't S3 show `s3://` or something like that on URLs?

**Comment by chriscct7** [comment]
Yeah but how many users will know to look for that in a URL? Plus it's not as easy to see.

The file repeat thing I agree with in premise though you could argue the whole box is supposed to be information about the file, so you could easily remove the word "file" from every box except the title of the metabox itself. The prefix of "file" repetition in this case I don't think really detracts from the user experience. It makes it easier to quickly scan and find items, versus looking for the "file size" and seeing "size" and then having to back-inference it to meaning "file size"

**Comment by cklosowski** [comment]
@mindctrl no, not all S3 files show the `s3://` handler. Some simply are just `/` for the bucket location.

**Comment by chriscct7** [comment]
@cklosowski he's referring to the download link URL for the download action, not the textbox where the file url goes

**Comment by cklosowski** [comment]
And as stated above, the file location _may_ be served from the local site, but might not be a fully qualified URL. It could simply be `/var/www/path/to/file.txt` or `/wp-content/uploads/file.zip`, which don't reference the actual place it's being served from by the normal user.

**Comment by cklosowski** [comment]
@chriscct7 I understand, but that's why file location can be useful.

**Comment by mindctrl** [comment]
It seems like it's solving a problem caused by the URL field being way too short to show anything useful about the file already uploaded and selected.

**Comment by cklosowski** [comment]
@mindctrl not true, again, files that are not fully qualified URLs, no matter how large the textbox is, can be confusing to a normal user on where they are being delivered from, since there is no reference to the location of the file in reference to the site.

**Comment by cklosowski** [comment]
For instance, here is an S3 file (need to work out getting the file information for it still on S3), that the URL section is pretty confusing for an end user who doesn't know how S3 works on a technical URL level.
![screen shot 2017-08-28 at 12 22 05 pm](https://user-images.githubusercontent.com/1390124/29798276-f1277bc2-8c10-11e7-89cc-7d54143ce8ab.png)

**Comment by cklosowski** [comment]
As a note, i'm still working out the kinks with the file storage solutions as far as delivery and file information.

**Comment by chriscct7** [comment]
But what user, outside of like Amazon s3, actually use a non-qualified URL and then do not realize it will be served from the same domain? I feel like Amazon s3 should just indicate the file is on AWS and then remove the file location thing and call it a day. Frankly the file type and file size probably aren't useful, because if the file url was longer, then you'd know what file type it is, and the file size is extraneous to most users.

2 alternative ideas:
No minimized area
![image](https://user-images.githubusercontent.com/1324637/29798343-85999682-8c2a-11e7-8c87-499470e6872b.png)
Minimized Area
![image](https://user-images.githubusercontent.com/1324637/29798347-8a3bfb80-8c2a-11e7-8ff0-88c03cc4e980.png)

**Comment by chriscct7** [comment]
Or you could do something like this:
![image](https://user-images.githubusercontent.com/1324637/29798444-f72c0654-8c2a-11e7-9b02-8a7cc8ca9612.png)

**Comment by cklosowski** [comment]
The file size was something born out of FES for store owners to download
vendor files. And since they did not upload it might not know how large the
file is, so it's a nice heads up to know before starting a large download
that might be done via mobile or a limited bandwidth connection.

>

**Comment by chriscct7** [comment]
So, we're thinking this in the EDD chat atm:
![image](https://user-images.githubusercontent.com/1324637/29798777-30a5276a-8c2d-11e7-901b-f35e2464ee40.png)

File URLs are the only thing on non-variations that people regularly edit as a typical user. If variations we add variation selection up there.
All other information moves to the sub-option (minimized).
We add the download link below the file URL if the file is added.

**Comment by SDavisMedia** [comment]
I'm 100% with @chriscct7's last reply here.

**Comment by cklosowski** [comment]
I can work that up.

**Comment by cklosowski** [comment]
This will change the JS a bit, since right now our expandable repeatables remove the entire additional section when you add a new row, but since the file name field needs to persist we'll just have to blank the rest of the fields.

**Comment by cklosowski** [comment]
Ok @SDavisMedia @chriscct7 I think this is meeting what you need.

Keep in mind, if the request for an external file returns a `WP_Error`, we'll actually just die out of the process of getting the mime and file size.

**Comment by cklosowski** [comment]
<img width="1194" alt="screen shot 2017-09-26 at 9 59 39 am" src="https://user-images.githubusercontent.com/1390124/30864409-80651726-a2a1-11e7-8677-347351e21996.png">

<img width="1195" alt="screen shot 2017-09-26 at 9 59 54 am" src="https://user-images.githubusercontent.com/1390124/30864422-8539461e-a2a1-11e7-9b19-b69ad6842fee.png">

**Comment by cklosowski** [comment]
Also a note, this *is* creating a file download log entry for when an admin downloads the file:
<img width="1473" alt="screen shot 2017-09-26 at 10 01 45 am" src="https://user-images.githubusercontent.com/1390124/30864503-c9ab1232-a2a1-11e7-8a16-fd5468186c8f.png">

**Comment by pippinsplugins** [comment]
That is looking really, really good!

**Comment by sunnyratilal** [comment]
Tested and working perfectly. Nice work @cklosowski!

**Comment by SDavisMedia** [comment]
I love it, @cklosowski. I think you nailed it.

Only issue I'm having is with the "Download File" link. I know you said it won't work for extension files. But I didn't see any mention of known issues with local files. When I click the link to download a locally hosted file, I get the following errors:

```

Notice: Undefined index: payment in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 916

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 919

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 920

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 921

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 922

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 923

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 979

Warning: Cannot modify header information - headers already sent by (output started at /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php:916) in /Applications/MAMP/htdocs/edd/wp-content/plugins/easy-digital-downloads/includes/process-download.php on line 613
```

Download log entries are still recorded, though.

**Comment by cklosowski** [comment]
Thanks @SDavisMedia I'll check that over, and I do want to support the external files, just need to do some more work on it. Wanted to get the :+1: from the interface before I started digging further in.

**Comment by pippinsplugins** [comment]
PR is here: https://github.com/easydigitaldownloads/easy-digital-downloads/pull/5979

**Comment by pippinsplugins** [comment]
I have resolved the merge conflicts here and brought `issue/3002` up to date with `master`. The merge conflict was nasty enough, however, that I had to abandon even trying to resolve the conflict with `process-download.php`, so any changes made there in `master` were lost in `issue/3002`. We'll need to bring those changes back over.

Those changes can be found here: https://github.com/easydigitaldownloads/easy-digital-downloads/commits/master/includes/process-download.php

We'll need to go back to at least Oct 1, possibly further.

**Comment by cklosowski** [comment]
We'll also need to look at how we can get 3rd party file handlers built into this. That was the final thing needed that I was struggling to get working.
